Gtk: change title prompt

b6k-dev opened the pull request:
This PR implements the title change functionality for the GTK app and closes https://github.com/ghostty-org/ghostty/issues/5769

Demo:

https://github.com/user-attachments/assets/cad611b3-b7bf-40ac-8d0f-11d2095fe525

For now I came up with a basic UI that i believe is sufficient, but I'm open to feedback and happy to iterate on it further. 
https://github.com/ghostty-org/ghostty/issues/5769#issuecomment-2660341107 - Regarding Mitchell's comment I checked Gnome Console and I could not find a similar feature.

b6k-dev committed:
Implement a prompt that allows the user to set the title

b6k-dev committed:
make the change of the title persistent & allow the user to restore to a default one

b6k-dev committed:
fix the edge case when user tries to revert the title to default and hasn't set a title before

b6k-dev committed:
remove outdated comment

b6k-dev committed:
free the terminal title when destroy is run

b6k-dev committed:
set the ok widget to be activated by default

jcollie commented on src/apprt/gtk/Surface.zig:
```diff
@@ -346,6 +346,11 @@ cursor: ?*c.GdkCursor = null,
 /// pass it to GTK.
 title_text: ?[:0]const u8 = null,
 
+/// The title of the surface as reported by the terminal.
```
See my comment here: https://github.com/ghostty-org/ghostty/issues/5769#issuecomment-2672676710

But I think we should not be changing the _surface_ title. This is about changing the _tab_ title. If we change the surface title the tab title will change every time that you switch focus to a different split if you have multiple splits in a tab.

jcollie replied:
After discussing this with Mitchell on Discord that's how he wants the feature to work right now. The current implementation seems a little awkward to me but maybe I'd have to take another look. So ignore this comment for now.

b6k-dev committed:
code review:
- use blueprint for the dialog content
- use zig-gobject bindings
- make the enum values lowercase

maciekbartczak commented on src/apprt/gtk/Surface.zig:
```diff
@@ -978,6 +978,31 @@ pub fn getTitle(self: *Surface) ?[:0]const u8 {
+    const context = try self.app.core_app.alloc.create(PromptTitleDialogContext);
+    context.self = self;
+
+    const dialog = c.gtk_message_dialog_new(window.window, c.GTK_DIALOG_MODAL, c.GTK_MESSAGE_QUESTION, c.GTK_BUTTONS_OK_CANCEL, "Set Tab Title");
```
Moved the dialog structure to a blueprint file.

maciekbartczak commented on src/apprt/gtk/Surface.zig:
```diff
@@ -969,15 +989,45 @@ fn updateTitleTimerExpired(ctx: ?*anyopaque) callconv(.C) c.gboolean {
+    defer builder.deinit();
+
+    const dialog: *adw.AlertDialog = @ptrCast(builder.getObject("prompt_title_dialog"));
+    dialog.addResponse("cancel", "Cancel");
+    dialog.addResponse("ok", "OK");
```
This could possibly be done from the `.blp` since there is a responses property, however I could not get this to work. Also there are no examples of how do define the responses from the blueprint, so I'm just leaving the definitions in the code right now.

pluiedev replied:
https://jwestman.pages.gitlab.gnome.org/blueprint-compiler/reference/extensions.html#adw-alertdialog-responses

b6k-dev committed:
code review:
- remove the menu entry defined in code

b6k-dev committed:
- remove the unused dialog context struct
- set the current title in the input buffer
- fix formatting

jcollie commented on src/apprt/gtk/Surface.zig:
```diff
@@ -1013,27 +1018,14 @@ const PromptTitleDialogContext = struct {
-    _ = c.g_signal_connect_data(dialog, "response", c.G_CALLBACK(&gtkPromptTitleResponse), context, null, c.G_CONNECT_DEFAULT);
-
-    c.gtk_widget_show(dialog);
+    dialog.choose(@ptrCast(window.window), null, @ptrCast(&gtkPromptTitleResponse), self);
```
Just a quick note since I don't have time for a full review right now, but the `@ptrCast` on the callback shouldn't be necessary using zig-gobject.

maciekbartczak replied:
Fixed.

b6k-dev committed:
code review:
- remove unnecessary @ptrCast
- set the default focus to the text entry field

maciekbartczak commented:
Two things that I modified from the original dialog version:
* the input box has focus by default
* the ok response has `suggested` apperance

![image](https://github.com/user-attachments/assets/61b5c6da-c912-4cc0-90c9-d1c98e60bda1)

pluiedev commented on src/apprt/gtk/Surface.zig:
```diff
@@ -2298,3 +2348,38 @@ fn g_value_holds(value_: ?*c.GValue, g_type: c.GType) bool {
+    const self = userdataSelf(ud.?);
+
+    const response = dialog.chooseFinish(result);
+    if (glib.strEqual("ok", response) != 0) {
```
Why not std.mem.orderZ(.., ..) == .eq?

maciekbartczak replied:
Fixed.

pluiedev commented on src/apprt/gtk/Surface.zig:
```diff
@@ -2298,3 +2348,38 @@ fn g_value_holds(value_: ?*c.GValue, g_type: c.GType) bool {
+
+    const response = dialog.chooseFinish(result);
+    if (glib.strEqual("ok", response) != 0) {
+        const title_entry: *gtk.Entry = @ptrCast(dialog.getExtraChild());
```
Use gobject.ext.cast

pluiedev commented on src/apprt/gtk/Surface.zig:
```diff
@@ -2298,3 +2348,38 @@ fn g_value_holds(value_: ?*c.GValue, g_type: c.GType) bool {
+        if (title.len == 0 and self.title_from_terminal != null) {
+            if (self.getTerminalTitle()) |terminal_title| {
+                self.setTitle(terminal_title, .user) catch {};
+                self.app.core_app.alloc.free(self.title_from_terminal.?);
```
The .? here feels like a bit of a smell... I think what I would do is to wrap this inside an if-capture

maciekbartczak replied:
I guess since getTerminalTitle already checks for the null this top level != null is redundant, so the simplest solution for me is to just remove it.

pluiedev replied:
Yea sure

pluiedev commented on src/apprt/gtk/Surface.zig:
```diff
@@ -2298,3 +2348,38 @@ fn g_value_holds(value_: ?*c.GValue, g_type: c.GType) bool {
+        // if the new title is empty and the user has set the title previously, restore the terminal provided title
+        if (title.len == 0 and self.title_from_terminal != null) {
+            if (self.getTerminalTitle()) |terminal_title| {
+                self.setTitle(terminal_title, .user) catch {};
```
Log error

maciekbartczak replied:
Added a log statement.

pluiedev commented on src/apprt/gtk/ui/prompt-title-dialog.blp:
```diff
@@ -0,0 +1,10 @@
+using Adw 1;
+
+Adw.AlertDialog prompt_title_dialog {
+  heading: "Change Terminal Title";
+  body: "Leave blank to restore the default title.";
```
Prefer [translatable strings](https://jwestman.pages.gitlab.gnome.org/blueprint-compiler/reference/values.html#translated-strings) for user-facing text

b6k-dev committed:
code review:
- move responses definition to the blueprint, use translatable strings
- minor changes in the response callback

b6k-dev committed:
format the blueprint file using `blueprint-compiler format`

pluiedev commented on src/apprt/gtk/Surface.zig:
```diff
@@ -969,15 +989,40 @@ fn updateTitleTimerExpired(ctx: ?*anyopaque) callconv(.C) c.gboolean {
+    var builder = Builder.init("prompt-title-dialog", .blp);
+    defer builder.deinit();
+
+    const entry: *gtk.Entry = @ptrCast(builder.getObject("title_entry"));
```
Never ptrCast between GObject objects - always do gobject.ext.cast

maciekbartczak replied:
fixed in https://github.com/ghostty-org/ghostty/pull/5905/commits/34c347dbbb4d77f89f2b89cdea8e59d9bd72cdc4

pluiedev commented on src/apprt/gtk/Surface.zig:
```diff
@@ -2298,3 +2343,39 @@ fn g_value_holds(value_: ?*c.GValue, g_type: c.GType) bool {
+
+    const response = dialog.chooseFinish(result);
+    if (std.mem.orderZ(u8, "ok", response) == .eq) {
+        const title_entry: *gtk.Entry = gobject.ext.cast(gtk.Entry, dialog.getExtraChild().?).?;
```
The explicit type annotation isn't necessary

maciekbartczak replied:
fixed in https://github.com/ghostty-org/ghostty/pull/5905/commits/34c347dbbb4d77f89f2b89cdea8e59d9bd72cdc4

pluiedev commented on src/apprt/gtk/Surface.zig:
```diff
@@ -2298,3 +2343,39 @@ fn g_value_holds(value_: ?*c.GValue, g_type: c.GType) bool {
+        if (title.len == 0) {
+            if (self.getTerminalTitle()) |terminal_title| {
+                self.setTitle(terminal_title, .user) catch |err| {
+                    log.err("Failed to set title: {}", .{err});
```
This isn't really *formalized* anywhere but our log messages usually look like
```suggestion
                    log.err("failed to set title={}", .{err});
```

maciekbartczak replied:
Fixed all log messages.

jcollie committed:
gtk: get 'Change Title' working with older distros

jcollie committed:
gtk: fix typos

jcollie committed:
snap: add blueprint-compiler to snap

jcollie commented:
@mitchellh @kenvandine since I had to touch the snap build I'd appreciate a quick once-over.

jcollie commented:
@maciekbartczak Thanks for this work! I added a few changes to get this compiling on older distros.

jcollie committed:
snap: add libxml2-utils (for xmllint) to snap

b6k-dev committed:
code review: fix log format

jcollie reviewed (approved):
OK, I think that this is ready to go!

pluiedev reviewed (approved):
The diff is a bit noisy due to all the blueprint fixes (frankly I think it would be better done in a separate PR) but the die's already been cast. Serve it forth

pluiedev commented:
Oh, merge conflicts :p

jcollie committed:
gtk: fix Builder api changes
